src/stochastic_search_iterator.jl

```diff
@@ -0,0 +1,167 @@
+    current_cost = calculate_cost(sampled_program, iter.cost_function, iter.examples, iter.grammar)
+    return (sampled_program, IteratorState(
+        current_program=sampled_program,
+        current_temperature=iter.initial_temperature,
+        best_program=sampled_program,
+        best_program_cost=current_cost))
+end
+
+
+"""
+The algorithm that constructs the iterator of StochasticSearchEnumerator. It has the following structure:
+
+1. get a random node location -> location,dict = neighbourhood(current_program)
+2. call propose on the current program getting a list of possbile replacements in the node location 
+3. iterate through all the possible replacements and perform the replacement in the current program 
+    4.  accept the new program by modifying the next_program or reject the new program
+5. return the new next_program
+"""
+function Base.iterate(iter::StochasticSearchEnumerator, current_state::IteratorState)
+    grammar, examples = iter.grammar, iter.examples
+    current_program = current_state.current_program
+    
+    current_cost = calculate_cost(current_program, iter.cost_function, examples, grammar)
+
+    new_temperature = iter.temperature(current_state.current_temperature)
+
+    # get the neighbour node location 
+    neighbourhood_node_location, dict = iter.neighbourhood(current_state.current_program, grammar)
+
+    # get the subprogram pointed by node-location
+    subprogram = get(current_program, neighbourhood_node_location)
+
+
+    @info "Start: $(rulenode2expr(current_program, grammar)), subexpr: $(rulenode2expr(subprogram, grammar)), cost: $current_cost
+            temp $new_temperature"
+
+    # propose new programs to consider. They are programs to put in the place of the node
+    possible_replacements = iter.propose(current_program, neighbourhood_node_location, grammar, iter.max_depth, dict)
+    
+    # the next program in the iteration
+    next_program = deepcopy(current_program)
+    possible_program = current_program
+    best_replacement = nothing
+    for possible_replacement in possible_replacements
+        # replace node at node_location with new_random 
+        if neighbourhood_node_location.i == 0
+            possible_program = possible_replacement
+        else
+            # update current_program with the subprogram generated
+            # this line mutates also the current_program. That is why we deepcopy at 115
+            neighbourhood_node_location.parent.children[neighbourhood_node_location.i] = possible_replacement
+        end
+        program_cost = calculate_cost(possible_program, iter.cost_function, examples, grammar)
+        if iter.accept(current_cost, program_cost, new_temperature) 
+            next_program = deepcopy(possible_program)
+            current_cost = program_cost
+            best_replacement = deepcopy(possible_replacement)
+        end
+    end
+
+    if current_cost < current_state.best_program_cost
```

Comment:
a shorter version

```julia
take_new = current_cost < current_state.nest_program_cost
next_state = IteratorState(
      current_program=next_program
      current_temperature=new_temperature
      best_program= take_new ? next_program : current_state.best_program
      best_program_cost= take_new ? current_cost : current_state.best_program_cost
)
```

Reply:
We are not tracking the `best_program` and `best_program_cost` anymore because `search_best` tracks  this properties.